tests/std/tests/P0401R6_allocate_at_least/test.cpp

```diff
@@ -0,0 +1,93 @@
+// Copyright (c) Microsoft Corporation.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include <algorithm>
+#include <cassert>
+#include <memory>
+#include <numeric>
+#include <type_traits>
+
+using namespace std;
+
+static_assert(is_standard_layout_v<allocation_result<int*>>);
+static_assert(is_trivially_copyable_v<allocation_result<int*>>);
+
+struct generous_allocator {
+    using value_type = int;
+
+    allocator<int> inner;
+
+    [[nodiscard]] _CONSTEXPR20_DYNALLOC int* allocate(const size_t count) {
+        return inner.allocate(max(count, static_cast<size_t>(16)));
+    }
+
+    [[nodiscard]] _CONSTEXPR20_DYNALLOC allocation_result<int*> allocate_at_least(const size_t count) {
+        return {inner.allocate(max(count, static_cast<size_t>(16))), max(count, static_cast<size_t>(16))};
+    }
+
+    _CONSTEXPR20_DYNALLOC void deallocate(int* ptr, const size_t count) {
+        inner.deallocate(ptr, max(count, static_cast<size_t>(16)));
+    }
+};
+
+struct strict_allocator {
+    using value_type = int;
+
+    allocator<int> inner;
+
+    [[nodiscard]] _CONSTEXPR20_DYNALLOC int* allocate(const size_t count) {
+        return inner.allocate(count);
+    }
+
+    _CONSTEXPR20_DYNALLOC void deallocate(int* ptr, const size_t count) {
+        inner.deallocate(ptr, count);
+    }
+};
+
+_CONSTEXPR20_DYNALLOC bool test() {
+    {
+        allocator<int> al;
+
+        same_as<allocation_result<int*>> auto result = al.allocate_at_least(5);
+        assert(result.ptr);
+        assert(result.count >= 5);
+        al.deallocate(result.ptr, 5);
+
+        result = al.allocate_at_least(7);
+        assert(result.ptr);
+        assert(result.count >= 7);
+        al.deallocate(result.ptr, result.count);
+
+        result = al.allocate_at_least(9);
+        assert(result.ptr);
+        assert(result.count >= 9);
+        al.deallocate(result.ptr, midpoint(static_cast<size_t>(9), result.count));

```

Comment:
I do not get that line

Reply:
> If p is memory that was obtained by a call to allocate_at_least, let ret be the value returned and req be the value passed as the first argument to that call. p is equal to ret.ptr and n is a value such that req ≤ n ≤ ret.count

This tests that a value between the requested size and the returned size works.